libjcat/jcat-libcrypto-pkcs7-engine.c

```diff
@@ -0,0 +1,484 @@
+	g_autoptr(GBytes) cert_blob = NULL;
+	g_autoptr(X509) crt = NULL;
+
+	filename =
+	    g_build_filename(jcat_engine_get_keyring_path(engine), "pki", "client.pem", NULL);
+	cert_blob = jcat_get_contents_bytes(filename, error);
+	if (cert_blob == NULL)
+		return NULL;
+	crt = jcat_libcrypto_pkcs7_load_crt_from_blob_pem(cert_blob, error);
+	if (crt == NULL)
+		return NULL;
+
+	return jcat_libcrypto_pkcs7_engine_verify(engine, blob, blob_signature, crt, flags, error);
+}
+
+static JcatResult *
+jcat_libcrypto_pkcs7_engine_pubkey_verify(JcatEngine *engine,
+					  GBytes *blob,
+					  GBytes *blob_signature,
+					  JcatVerifyFlags flags,
+					  GError **error)
+{
+	return jcat_libcrypto_pkcs7_engine_verify(engine, blob, blob_signature, NULL, flags, error);
+}
+
+static JcatBlob *
+jcat_libcrypto_pkcs7_engine_pubkey_sign(JcatEngine *engine,
+					GBytes *blob,
+					GBytes *cert,
+					GBytes *privkey,
+					JcatSignFlags flags,
+					GError **error)
+{
+	g_autoptr(EVP_PKEY) key = NULL;
+	g_autoptr(X509) crt = NULL;
+	g_autoptr(CMS_ContentInfo) cms_ci = NULL;
+	g_autoptr(BIO) blob_bio = NULL;
+	g_autoptr(BIO) sig_bio = NULL;
+	g_autoptr(GString) rewrap = NULL;
+	gchar *bio_buf;
+	gsize bio_len;
+	guint signing_flags = CMS_BINARY | CMS_DETACHED | CMS_NOSMIMECAP;
+
+	/* nothing to do */
+	if (g_bytes_get_size(blob) == 0) {
+		g_set_error_literal(error, G_IO_ERROR, G_IO_ERROR_INVALID_ARGUMENT, "nothing to do");
+		return NULL;
+	}
+
+	/* load keys */
+	key = jcat_libcrypto_pkcs7_load_privkey_from_blob_pem(privkey, error);
+	if (key == NULL)
+		return NULL;
+	crt = jcat_libcrypto_pkcs7_load_crt_from_blob_pem(cert, error);
+	if (crt == NULL)
+		return NULL;
+
+	/* get the digest algorithm from the public key */
+	// TODO: The gnutls uses the keys "preferred_hash_algorithm". Port that if necissary.
+	// gchar mdname[256] = {0};
+	// gint default_digest_nid = 0;
+	// EVP_PKEY_get_default_digest_name(key, mdname, sizeof(mdname));
+	// EVP_PKEY_get_default_digest_nid(key, &default_digest_nid);
+	// g_debug("pubkey_sign pref digest %s %d", mdname, default_digest_nid);
+
+	/* sign data */
+	// TODO: Implement these sign ing flags
+	// TODO: CMS_NO_SIGNING_TIME was added in OpenSSL 3.5
+	// if (!(flags & JCAT_SIGN_FLAG_ADD_TIMESTAMP))
+	//	signing_flags |= CMS_NO_SIGNING_TIME;
+	if (!(flags & JCAT_SIGN_FLAG_ADD_CERT))
+		signing_flags |= CMS_NOCERTS;
+	blob_bio = BIO_new_mem_buf(g_bytes_get_data(blob, NULL), g_bytes_get_size(blob));
+	cms_ci = CMS_sign(crt, key, NULL, blob_bio, signing_flags);
+
+	sig_bio = BIO_new(BIO_s_mem());
+
+	if (!PEM_write_bio_CMS(sig_bio, cms_ci)) {
+		g_autofree gchar *error_str = jcat_libcrypto_get_errors();
+		g_set_error(error,
+			    G_IO_ERROR,
+			    G_IO_ERROR_INVALID_DATA,
+			    "failed to encode pkcs7 as pem: %s",
+			    error_str);
+		return NULL;
+	}
+
+	bio_len = BIO_get_mem_data(sig_bio, &bio_buf);
+
+	rewrap = g_string_new_len(bio_buf, bio_len);
+	g_string_replace(rewrap, " CMS-----", " PKCS7-----", 0);
```

Comment:
is there a way to coerce openssl to output a PKCS7 header? This smells like papering over a missing flag somewhere.

Reply:
I had a look through the chain of macros that defines `PEM_write_bio_CMS` and, yes, there is a slightly less readable way.
https://github.com/hughsie/libjcat/pull/177/commits/2301f08fc77b75d13143198442dbcb32f1a9a8ee